### PR TITLE
TEL-6005: Initial implementation of RTCP-XR

### PR DESCRIPTION
**Pull request by damirn**

### CODE REVIEW

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/switch_core_media.c [hunk 1]
```diff
@@ -12243,8 +12246,10 @@ SWITCH_DECLARE(void) switch_core_media_gen_local_sdp(switch_core_session_t *sess
 			if (a_engine->rtcp_mux > 0) {
 				switch_snprintf(buf + strlen(buf), SDPBUFLEN - strlen(buf), "a=rtcp-mux\r\n");
 				switch_snprintf(buf + strlen(buf), SDPBUFLEN - strlen(buf), "a=rtcp:%d IN %s %s\r\n", port, family, ip);
+				switch_snprintf(buf + strlen(buf), SDPBUFLEN - strlen(buf), "a=rtcp-xr:rcvr-rtt=all:10000 stat-summary=loss,dup,jitt,TTL voip-metrics\r\n");
```
[the remark this answers is not in the document]
Comment (damirn) [reply src/switch_core_media.c:12249]: Right, that was my initial idea, but I wanted to get this working ASAP, so it slipped my mind; will implement it.

File: src/switch_core_media.c [hunk 2]
```diff
@@ -12243,8 +12246,10 @@ SWITCH_DECLARE(void) switch_core_media_gen_local_sdp(switch_core_session_t *sess
...
 			} else {
 				switch_snprintf(buf + strlen(buf), SDPBUFLEN - strlen(buf), "a=rtcp:%d IN %s %s\r\n", port + 1, family, ip);
+				switch_snprintf(buf + strlen(buf), SDPBUFLEN - strlen(buf), "a=rtcp-xr:rcvr-rtt=all:10000 stat-summary=loss,dup,jitt,TTL voip-metrics\r\n");
```
[the remark this answers is not in the document]
Comment (damirn) [reply src/switch_core_media.c]: should be now configurable via sofia profile and / or channel variable

File: src/switch_core_media.c [hunk 3]
```diff
@@ -12248,6 +12252,10 @@ SWITCH_DECLARE(void) switch_core_media_gen_local_sdp(switch_core_session_t *sess
 			}
 		}
 
+		if (switch_channel_var_true(session->channel, "use_rtcp_xr") || (!switch_channel_var_false(session->channel, "use_rtcp_xr") && smh->mparams->offer_rtcp_xr)) {
```
Comment (dev-ryanc) [inline src/switch_core_media.c:12255]: Should this be inside the validation of `rtcp_audio_interval_msec`? If we disable rtcp then we shouldn't send or offer rtcp-xr.

File: src/switch_core_media.c [hunk 4]
```diff
@@ -11456,6 +11456,10 @@ static void generate_m(switch_core_session_t *session, char *buf, size_t buflen,
 		}
 	}
 
+	if (switch_channel_var_true(session->channel, "use_rtcp_xr") || (!switch_channel_var_false(session->channel, "use_rtcp_xr") && smh->mparams->offer_rtcp_xr)) {
```
Comment (dev-ryanc) [inline src/switch_core_media.c:11459]: Should this be inside the validation of rtcp_audio_interval_msec? If we disable rtcp then we shouldn't send or offer rtcp-xr.

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/switch_core_media.c (hunk 3 above)
Comment (damirn) [reply src/switch_core_media.c]: ok, makes sense

**Review by dev-ryanc — APPROVED** [review]

**Review by baloeng — APPROVED** [review]

**Pull request merged**